### DISCUSSION TITLE
async retry

### DISCUSSION THREAD

**Discussion by scr2em** [TanStack/query#2770] [Ideas]
HI, 

in my old project I was making an Axios interceptor and queuing all failed requests then fetch and refresh my `refreshToken` then retry all the failed requests. 
with React Query I don't need half of the setup i made before (thanks React Query <3) 
but now I'm stuck with the refreshToken mechanism as I want to move it to React Query instead of Axios Interceptor 

as the docs say in useQuery options

  ```
  retry: boolean | number | (failureCount: number, error: TError) => boolean
  ```
  
**I think it would be a good idea to use async function for retry** what do you think about this ?

```
  const queryClient = new QueryClient({
    defaultOptions: {
      queries: {
        // query options
        retry: async (count, error) => {
          const {status} =error.resposne
                if (status && [401, 403].includes(status)) {
                      return await refreshAuthToken()
                  }
          return false // and redirect to login
        },   
    },
  })
  ```

**Comment by sred907** [comment]
Hi i am also looking for this sort of behaviour. But I would like to handle this at a global level. Is it possible? 
eg: for any failed request I want to fetch refresh token and retry the failed call again.

**Comment by TkDodo** [comment]
I'm not opposed to this, just need to wrap my head around it completely.

1) You would need to return a `Promise<boolean>`, right? So the example would need to be something like:

```
        retry: async (count, error) => {
          const {status} = error.resposne
                if (status && [401, 403].includes(status)) {
                      await refreshAuthToken()
                      return true
                  }
          return false
        }, 
```

2) what is `await refreshAuthToken` doing? Is it writing the auth token somewhere where the `queryFn` will then pick it up?
3) what happens if the `refreshAuthToken` request fails? We would get a rejected promise, which we would likely interpret as "no retry" ?
4) currently, you can do all of this inside the queryFn. Is that so bad?

```
useQuery(key, async () => {
  try {
    const { data } = await axios.get(...)
    return data
  } catch(error) {
    const {status} =error.response
    if (status && [401, 403].includes(status)) {
      const token = await refreshAuthToken()
      const { data } = await axios.get(token, ...)
      return data
    }
    throw error
  }
})
```
I get that if we want this globally, we'd need to use the defaultQueryFn.

**Reply by sred907** [reply]
For 4. currently, you can do all of this inside the queryFn. Is that so bad? - In that case we will have to do it across the whole application. For now I have setup an axios interceptor at global level to achieve this. It would be great if we could somehow do it in defaultQueryFn

**Reply by tom-sherman** [reply]
We solve this by creating API services that use an underlying network service that handles retries, auth, logging, etc. We then call the API services in queryFn. This is a much more explicitly and straightforward seperation of concerns.

IMO this feature doesn't sit with react-query, it's an implementation detail of your query functions.

**Reply by bennettdams** [reply]
I'm also looking for a solution to this. 


1. Maybe I'm missing something, but using the `defaultQueryFn` doesn't make sense, because not every request is supposed to get new refresh tokens.
2. I also thought about using the [global error handler](https://tkdodo.eu/blog/react-query-error-handling#the-global-callbacks) and fetch refresh tokens there - but this only works if there's only **one** ongoing (failing 401) request. As soon as e.g. you execute two requests on page load that both produce 401 responses, you would refetch refresh tokens twice in the `onError` callback.
    2.1. One could save whether there is an ongoing refresh request globally and only execute the first occuring request. But now the second request that failed does not know that the first request already fetched refresh tokens and fails.

Handling refresh tokens by just refreshing them in an interval is easy, but what is the canonical way to get refresh tokens when you only refetch them on a 401 response? I'm asking this specifically, as it is considered the [standard for the OAuth flow](https://www.oauth.com/oauth2-servers/making-authenticated-requests/refreshing-an-access-token/):

> If you make an API request and the token has expired already, you’ll get back a response indicating as such. You can check for this specific error message, and then refresh the token and try the request again. [...] When your application recognizes this specific error, it can then make a request to the token endpoint using the refresh token it previously received, and will get back a new access token it can use to retry the original request.

Update:

Regarding refetching failed requests: You can combine refreshing the token with invalidating all active queries. That way only one is fetching new tokens, but all other queries are re-fetched afterwards:

```ts
/**
 * It is possible that we execute multiple backend requests at the same time that both respond with "unauthenticated" status.
 * With this variable we make sure that we don't execute multiple "refresh" requests, one is enough.
 */
let hasOngoingRefreshRequest = false

export const queryClient = new QueryClient({
  defaultOptions: {
    queries: {
      retry: 3,
      ...
    },
  },
  queryCache: new QueryCache({
    /*
     * Global error callback for all queries.
     * We use this to react on a 401 response, where we want to try to refresh our tokens via the refresh token.
     */
    onError: async (error) => {
      if (error === errorMessageUnauthenticated && !hasOngoingRefreshRequest) {
        hasOngoingRefreshRequest = true

        try {
          const refreshResponse = await apiRefreshToken()

          if (refreshResponse.ok) {
            console.info('Renewing request was successful.')

            // we invalidate all active queries to let them refetch automatically
            queryClient.invalidateQueries({ active: true })

            hasOngoingRefreshRequest = false
          }
        } catch (error) {
          hasOngoingRefreshRequest = false

          console.error('Error while renewing tokens:', error)

          // if refresh request fails, we redirect to sign-in
          Router.push(ROUTES.SIGNIN)
        }
      }
    },
  }),
})
```
The only thing missing here is being able to say to all queries that they shouldn't retry if it is a 401 response. I have a global retry of "3", but I would need "0" if it is a 401 error, as I just want to execute refreshing tokens in that scenario.

**Reply by tom-sherman** [reply]
You're introducing a bunch of complexity by making React Query handle retries for 401 errors. As in my above comment I'd recommend simplifying by handling 401s yourself inside the `queryFn`.

Conceptually 401 errors are part of your API so they should be handled at the API level (the queryFn) not at the caching level (React Query).

**Reply by bennettdams** [reply]
> handling 401s yourself inside the queryFn

One would need to do this in every single query, which is not something I would like to do 😅

> should be handled at the API level (the queryFn) not at the caching level (React Query).

We solely interact with the API via React Query, so in my opinion there's no point in introducing another layer on top/inbetween. The only exception is fetching data from the server (Next.js), but these errors are not handled by React Query anyways.

**Reply by TkDodo** [reply]
> We solely interact with the API via React Query, so in my opinion there's no point in introducing another layer on top/inbetween.

most people use some sort of layer in between making api requests. for example, if you use `axios` or `ky` (based on `fetch`), you already have good places to hook into (interceptors / hooks).

I somewhow doubt that you're calling `fetch()` in every queryFn, because that means you have to check for `!response.ok` to throw errors, do json parsing, set correct headers etc everywhere! So even if you don't use a lib, you might have a separate abstraction over plain `fetch` - so that is where you can handle this.

---

However, I also think that async retries might be a good idea and a fine case to handle this. Maybe we should try to find other use-cases for async retries that are not about 401 errors? For example, we have an async toast api, which means we could ask the user: "Something failed, do you want to retry?" in the ui _before_ doing the retry...

**Reply by bennettdams** [reply]
> I somewhow doubt that you're calling fetch() in every queryFn

Correct, I have abstractions for `fetch` itself and each API call has it's own helper fn to execute the wrapper fetch fn.

> so that is where you can handle this.

Is there an example of what that might look like? I guess it is easy to catch the 401 response, but you'd e.g. need to refetch whatever initially wanted to fetch, so you need to _remember_ the callers and so on. I just took the easy route with invalidating queries, but I'd be glad about some kind of example to handle this use case the right way.

**Reply by TkDodo** [reply]
```
const myFetch = async (url, options) => {
  const response = await fetch(url, options)
  if (!response.ok) {
    if (response.status === 401) {
      await refetchAuthToken()

      return myFetch(url, options)
    }
    throw new Error()
  }
  return response.json()
}
```

this is a bit pseudo code, I would likely extract fetch + ok checking and json parsing into a separate function instead of calling `myFetch` recursively as that can lead to infinite errors if you get 401s indefinitely :)

**Reply by bennettdams** [reply]
That's actually how I started implementing this (even before the solution with RQ's `onError` mentioned above), but this has one major flaw: Imagine you're executing multiple requests on page load. Getting new auth tokens on 401 should only be done for _one_ of the requests (based on the OAuth flow). So maybe you save the ongoing `refetchAuthToken` request for the first one (see `hasOngoingRefreshRequest` above), but now all other requests have to wait for a successful `refetchAuthToken` response of whoever executed it first and only _then_ retry their original request, maybe via a central place to store them first. But you'd need to build that - using `onError` with `invalidateQueries` seemed way easier.

**Reply by TkDodo** [reply]
async retries wouldn't support this either because retries run per query and obtaining a new auth token is likely a mutation that isn't automatically deduped ...

**Reply by bennettdams** [reply]
Mhh, true, you would need to be able to wait in the other queries for a successful response of the `refetchAuthToken` query, otherwise refetching them wouldn't make sense.
Actually, something like a callback `onSuccessOfOtherQuery: (queries) => ...` would help, where all queries just retry when the refetch query is 200. But we're now far off from the original thread 😅 I'm still happy with the `onError` solution above, maybe I don't need async retries.

**Reply by damianbeles** [reply]
@bennettdams how did you solve the problem?

**Reply by bennettdams** [reply]
@damianbeles I'm still happy with [my mentioned solution above](https://github.com/TanStack/query/discussions/2770#discussioncomment-3584906).

**Reply by ammar-madni** [reply]
@tom-sherman Am I correct in assuming that you don't make the request to refresh your access token via a tanstack-query mutation? My setup sounds a lot like what you've suggested but I've not found a way to make it work with a mutation, this is because as soon as I wrap it in a mutation I can't await for the data from the response and then use the access token from the response to retry the original request that failed.

**Reply by tom-sherman** [reply]
> Am I correct in assuming that you don't make the request to refresh your access token via a tanstack-query mutation?

Yes, that's right. We have a separate service that manages the token lifetime in memory. There might be a way to move this into RQ, but I could never figure it out.

**Reply by yeonah98** [reply]
Hello @TkDodo,

I've been exploring error handling mechanisms in the context of using React Query (tanStack Query) alongside Axios, and I'm particularly interested in understanding the best practices for utilizing React Query's onError callback in various scenarios.

When dealing with a 401 error, for instance, I understand that Axios interceptors are commonly used to catch and handle these errors globally. However, I'm curious about the specific cases where React Query's onError should be leveraged instead. Could you provide insight into when it's more appropriate to use onError within React Query for error handling, especially in situations that might not be adequately addressed by Axios interceptors?

Additionally, could you share examples of errors or situations during the fetching or mutation processes that are not caught by Axios interceptors and how React Query's onError can be effectively used to handle these cases? I'm looking to gain a deeper understanding of how to distinguish between errors that should be handled by Axios interceptors versus those that should be managed within the React Query framework.

Thank you for your guidance and insights😀 

(with @okhan1998)

**Comment by arnaudbzn** [comment]
The approach is interesting to get rid of some logic inside the `AuthenticationProvider` I use to create to manage this problem.
To refresh tokens, I check access token expiration to be more precise about the need to refresh the token because a 401 can also mean that the access token is invalid or missing.
A 403 is an authorization error so it should not be related to a token expiration.

**Comment by arnaudbzn** [comment]
There is also a "proactive" solution if your access token is a JWT:
In your queryFn (your async fetcher), before calling your API, you can [decode](https://github.com/auth0/jwt-decode) the jwt access token and check the `exp` claim.
So you can do `const token = await refreshAuthToken()` if the access token has expired (or soon to be expired).
You can even optimize the solution by storing the `exp` value  in the query `meta`: so no need to decode the access token for each query.

**Comment by mwmcode** [comment]
Hi @TkDodo 

Current retry function logic `(typeof retry === 'function' && retry(failureCount, error))`  will always evaluate to `true` if we pass `async` functions.



Can we make it so it throws (or at least fails) if retry is a function that does **not** return a `boolean`?

I was debugging why it wasn't acting as expected then came to realise that it doesn't support `async` fns.

```js
// packages/query-core/src/retryer.ts
//...
        // Do we need to retry the request?
        const retry = config.retry ?? 3
        const retryDelay = config.retryDelay ?? defaultRetryDelay
        const delay =
          typeof retryDelay === 'function'
            ? retryDelay(failureCount, error)
            : retryDelay
        const shouldRetry =
          retry === true ||
          (typeof retry === 'number' && failureCount < retry) ||
          (typeof retry === 'function' && retry(failureCount, error))
```
can we change the last line to
```js
(typeof retry === 'function' && retry(failureCount, error) === true)
```
> I think a more DX friendly approach would check return type & throw if not a boolean
>```js
> // ...check if retry true..number..
> // ... check if it's of type function then
>const retryFnResult = retry(failureCount, error);
>if (typeof retryFnResult !== 'boolean') throw new Error("retry function must return a boolean")
>```

**Reply by TkDodo** [reply]
> Can we make it so it throws (or at least fails) if retry is a function that does not return a boolean?

it should fail with typescript because it has to return a `boolean` right now, not a `Promise<boolean>`.

> can we change the last line to

I'm reluctant to add runtime checks when things are already covered by types and documentation, sorry. Otherwise, everything would be full of null / undefined / NaN runtime checks ...

**Reply by mwmcode** [reply]
Agree 100%. 

Can you elaborate how TS would catch this? My linter (screenshot) is not complaining when I pass `async` function. _am I missing something here?_

![Screen Shot 2022-10-28 at 5 17 19 PM](https://user-images.githubusercontent.com/13101565/198734398-ee7be877-5528-4c39-9dd2-8d2cf08003cb.png)

**Reply by TkDodo** [reply]
It does error nicely for me with:

`Type 'Promise<boolean>' is not assignable to type 'boolean'.`

https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgVwM4FMCKz1QJ5wC+cAZlBCHAOQACMAhgHaoMDGA1gPRTr2swBaAI458VALAAoKWiyi8ACgDaVEhAhUAugBo4CgJRwAvAD44ABXIhgGAHQ9UEADYA3dAtXqq+3Qilw4Hhh8AC44elQ8RlY9Q1NEfwDA9BhkKEYLKxt0e3RHV3dgnH1EwilCfSA

**Reply by mwmcode** [reply]
Thanks @TkDodo 
It must be an issue with my setup then.

**Reply by mwmcode** [reply]
> This is getting spammy, sorry.

Just wanted to point out that it seems NOT to complain when you call `useQuery` with an object.

```js
const result = useQuery({
  queryKey: ['foo'],
  queryFn: () => Promise.resolve('foo'),
  retry: async () => {
     Promise.resolve(true)
  },
})
```
https://www.typescriptlang.org/play?ssl=3&ssc=1&pln=4&pc=1#code/JYWwDg9gTgLgBAbzgVwM4FMCKz1QJ5wC+cAZlBCHAOQACMAhgHaoMDGA1gPRTr2swBaAI458VALAAoKawjN4PVMgA28ALwoM2XHgAUCKXDgidAaXR4AXHADaVEhAhUAugBpDx0XgBija7oBKODUAPjgABXIQYAwAOkUIZQA3dF17RyoA90kjHhh8a3pUPEZWOEDgsIMcoyNIihj0ePRURJTdfJwAj0Jswm7pSSA


Edit: or rather gives some different errors! ...idk... TS 😖

**Reply by TkDodo** [reply]
yeah it does complain, just differently, and very weirdly so:

```
    Argument of type '{ queryKey: string[]; queryFn: () => Promise<string>; retry: () => Promise<void>; }' is not assignable to parameter of type 'QueryKey'.
      Object literal may only specify known properties, and 'queryKey' does not exist in type 'readonly unknown[]'.(2769)
```

but I think this is because of our overloads and we'll get rid of them in v5. if we remove the overloads, the error is much better, see [this playground](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgRQK4FMoE8DS6sA0cAqgM7pqZYBK6pqANjEWRRtgPJgzAQB2pIgBF0AM2B90AE1aVstekyJysAMVR8Axj35wAvnFFQIIOAHIAAjACGAm5oDWAeijpr2gLQBHdljMBYACggqXRNBmtXQw1tXj44VHIVAB4guDgAFRVVPiFrGzgAXgS+Bz4IAHc+AjTMgFEoYygikrLK6tqMvILirN8c7usawPS+qjwsOHQADxh0PilSFF8JlpUJ4YA+AApaiG440gAuOA4QYBhUkfSSJN8uHQFksewB-KH6xugiLvef9fwm2GN3MEguwGsDEGAWumzgADJEHAwTxIYMAPwnbYASiKcI0oXEkik+mG2JOsl8CkYl1+Nh+DSamyCITCESiohijwSdyoV1G2Vy7xaGjaVWBnyaItK5XFnUGLReaiF9M6AMmMzmCyW6rWK3wW12132j2Op3Ol1q6UpVAeh2egsGDK+UB+Tsy6qBVtBfHBaPeMPScMRSBREKh7xOSreBQAPnBtji8R7+irrLi9GSTiIidIbfI6DTnu6MozoMzgoFQuFIuholpuYk2Hy1amFcVRbKOtdSy7pWLu6N2ymqDGhq3xvgprN5otlpPJsV1Ya9gd+Gb81g7euHW2-pLvpkS56s7dmwXFLSS2WoBWgpp1-BXJeRbzsNsELUfAuTgBtMyiBAEBmAAuhK36vHwWK4oUcIAArGOc5AAHTPhADAAG7oNsAFAWY2ISq4MDYCc1ikFgWgJjBcKftcNwISYwAoWhmHYcRGDYrUmZBHonGVkAA).

```
    Type '() => Promise<void>' is not assignable to type 'RetryValue<unknown> | undefined'.
      Type '() => Promise<void>' is not assignable to type 'ShouldRetryFunction<unknown>'.
        Type 'Promise<void>' is not assignable to type 'boolean'.
```

**Reply by mwmcode** [reply]
I see it now. Thanks again.

This looks neat 👍 straight to the point. Excited about v5 already :) 
![Screen Shot 2022-10-29 at 9 46 43 AM](https://user-images.githubusercontent.com/13101565/198835047-14e5a49c-b76b-458c-8710-aab359fcc5ee.png)

**Comment by darshanatnintee** [comment]
Hey @TkDodo I tried `async retry` to refresh my token, Its stuck in infinite loop. 

I did test this in `v4` and `rc`

```
const retry = (count, error) => {
 if (count > 3) {
    console.log('WE ARE DONE WITH TRYYYYY')
  
  // this is not working 
    return Promise.resolve(false)
  }

  const axiosError = error as AxiosError
  const codeStatus = axiosError?.response?.status || 0

  console.log('CODE STATUS', codeStatus)

  if (codeStatus === 401) {
    console.log('GENERATING TOKEN')
    return Promise.resolve(false)
  }
  return Promise.resolve(true)
}
}
```

**Reply by TkDodo** [reply]
that's because `retry` isn't async.

**Reply by darshanatnintee** [reply]
Okay, it doesn't expect promise? Can we run async logic inside `retry` function?

**Reply by TkDodo** [reply]
no and no.

**Reply by darshanatnintee** [reply]
can I expect an async nature in the future?

**Comment by maapteh** [comment]
I also see a need for this but then on the hook level as well (not generic client). The current retry inside a hooks itself is currently only a number. But in our case 400 errors for example have explanations with `key: FOO` and we then decide what to do, but on 500 errors there can be some other error which we then simply want to retry, or really specific keys in 400 responses. 

So my question basically is can we have a retry where we can act on the status code / error response.

**Reply by TkDodo** [reply]
you can already do that because reading a status code is not async ?

**Comment by fl0ydj** [comment]
We would need async retry for advanced error handling.
We are fetching data from the backend using useQuery and the data fetched influences code paths in useMutations. In case of stale data, this might lead to errors in the useMutation.

We are using websockets to invalidate queries to avoid staleness but in case of websocket timeouts, we want to have a backup invalidate mechanism if the usemutation throws specific error types. "Invalidation" here is not enough though and we actually need to await queryClient.refetchQueries which we can't do within the retry function.

We ended up wrapping the usemutation and usequery hook for now to add this to the query/mutation fn but its not ideal.
